bigchaindb/models.py

```diff
@@ -0,0 +1,209 @@
+
+        if not self.is_signature_valid():
+            raise InvalidSignature('Block signature invalid')
+
+        # Finally: Tentative assumption that every blockchain will want to
+        # validate all transactions in each block
+        for tx in self.transactions:
+            # NOTE: If a transaction is not valid, `is_valid` will throw an
+            #       an exception and block validation will be canceled.
+            bigchain.validate_transaction(tx)
+
+        return self
+
+    def sign(self, private_key):
+        block_body = self.to_dict()
+        block_serialized = serialize(block_body['block'])
+        private_key = SigningKey(private_key)
+        self.signature = private_key.sign(block_serialized)
+        return self
+
+    def is_signature_valid(self):
+        block = self.to_dict()['block']
+        block_serialized = serialize(block)
+        verifying_key = VerifyingKey(block['node_pubkey'])
+        try:
+            return verifying_key.verify(block_serialized, self.signature)
+        except (ValueError, AttributeError):
+            return False
+
+    @classmethod
+    def from_dict(cls, block_body):
+        block = block_body['block']
+        block_serialized = serialize(block)
+        block_id = hash_data(block_serialized)
+        verifying_key = VerifyingKey(block['node_pubkey'])
+
+        try:
+            signature = block_body['signature']
+        except KeyError:
+            signature = None
+
+        if block_id != block_body['id']:
+            raise InvalidHash()
+
+        if signature is not None:
+            # TODO FOR CC: `verify` should accept an arbitrary string and then
+            #              return `False` or `True`. Shouldn't throw value err.
+            try:
+                signature_valid = verifying_key.verify(block_serialized,
+                                                       signature)
+            except ValueError:
+                signature_valid = False
+            if signature_valid is False:
+                raise InvalidSignature('Invalid block signature')
+
+        transactions = [Transaction.from_dict(tx) for tx
+                        in block['transactions']]
+
+        return cls(transactions, block['node_pubkey'],
+                   block['timestamp'], block['voters'], signature)
+
+    @property
+    def id(self):
+        return self.to_dict()['id']
+
+    def to_dict(self):
+        if len(self.transactions) == 0:
+            raise OperationError('Empty block creation is not allowed')
+
+        block = {
+            'timestamp': self.timestamp,
+            'transactions': [tx.to_dict() for tx in self.transactions],
+            'node_pubkey': self.node_pubkey,
+            'voters': self.voters,
+        }
+        block_serialized = serialize(block)
+        block_id = hash_data(block_serialized)
+
+        return {
+            'id': block_id,
+            'block': block,
+            'signature': self.signature,
+        }
```

Comment:
In the future, it might be nice to cache this since this is potentially quite expensive (1000 `tx.to_dicts()`) to be used each time for random look ups, like `.id`.